### PR TITLE
Refactor handshake state machine

### LINKED ISSUE #143: Use single state machine when handling packets in `conn.go`
Credit to @ic0ns @pfg666 and @jurajsomorovsky for this report, thank you!

----

It is hard to analyze the code because we have so much state (and all the ways they interact with each other). We should introduce a single state machine that has a clear progression so that the code is cleaner and is easier to analyze.

### LINKED ISSUE #145: Don't decrypt packets without receiving a ChangeCipherSpec first
Credit to @ic0ns @pfg666 and @jurajsomorovsky for this report, thank you!

---

We should just discard all messages in epoch 1 until we have gotten a `ChangeCipherSpec`. 

The only time I could see this being an issue is if you had a `NULL` ciphersuite, since incoming packets aren't handled until we have local crypto state there shouldn't be any harm.

### LINKED ISSUE #178: TestPionE2ESimplePSK is flaky
On CI environment, `TestPionE2ESimplePSK` sometimes fail due to timeout.

```
=== RUN   TestPionE2ESimplePSK
goroutine profile: total 4
1 @ 0x457eef 0x42db56 0x42db2c 0x42d8b5 0x6b9e57 0x6b9872 0x485a61
#	0x6b9e56	github.com/pion/dtls/v2/internal/udp.(*Listener).getConn+0x326	/home/travis/gopath/src/github.com/pion/dtls/internal/udp/conn.go:144
#	0x6b9871	github.com/pion/dtls/v2/internal/udp.(*Listener).readLoop+0x171	/home/travis/gopath/src/github.com/pion/dtls/internal/udp/conn.go:119

1 @ 0x457eef 0x42e84f 0x42e825 0x42e5db 0x53ae12 0x540c89 0x53a6c4 0x53ccf4 0x53b56c 0x7411a5 0x457adc 0x485a61
#	0x53ae11	testing.(*T).Run+0x6d1		/home/travis/.gimme/versions/go1.12.linux.amd64/src/testing/testing.go:917
#	0x540c88	testing.runTests.func1+0xa8	/home/travis/.gimme/versions/go1.12.linux.amd64/src/testing/testing.go:1157
#	0x53a6c3	testing.tRunner+0x163		/home/travis/.gimme/versions/go1.12.linux.amd64/src/testing/testing.go:865
#	0x53ccf3	testing.runTests+0x523		/home/travis/.gimme/versions/go1.12.linux.amd64/src/testing/testing.go:1155
#	0x53b56b	testing.(*M).Run+0x2eb		/home/travis/.gimme/versions/go1.12.linux.amd64/src/testing/testing.go:1072
#	0x7411a4	main.main+0x344			_testmain.go:250
#	0x457adb	runtime.main+0x20b		/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/proc.go:200

1 @ 0x457eef 0x46922d 0x469203 0x468f9d 0x48ec19 0x48fe4a 0x6b8e98 0x71c417 0x73dfdd 0x73ac2c 0x73b2cd 0x53a6c4 0x485a61
#	0x468f9c	sync.runtime_SemacquireMutex+0x3c				/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/sema.go:71
#	0x48ec18	sync.(*Mutex).Lock+0x148					/home/travis/.gimme/versions/go1.12.linux.amd64/src/sync/mutex.go:134
#	0x48fe49	sync.(*RWMutex).Lock+0x49					/home/travis/.gimme/versions/go1.12.linux.amd64/src/sync/rwmutex.go:93
#	0x6b8e97	github.com/pion/dtls/v2/internal/udp.(*Listener).Close+0x67	/home/travis/gopath/src/github.com/pion/dtls/internal/udp/conn.go:45
#	0x71c416	github.com/pion/dtls/v2.(*Listener).Close+0x76			/home/travis/gopath/src/github.com/pion/dtls/listener.go:46
#	0x73dfdc	github.com/pion/dtls/v2/e2e.assertE2ECommunication.func3+0x1cc	/home/travis/gopath/src/github.com/pion/dtls/e2e/e2e_test.go:143
#	0x73ac2b	github.com/pion/dtls/v2/e2e.assertE2ECommunication+0x3db	/home/travis/gopath/src/github.com/pion/dtls/e2e/e2e_test.go:177
#	0x73b2cc	github.com/pion/dtls/v2/e2e.TestPionE2ESimplePSK+0xfc		/home/travis/gopath/src/github.com/pion/dtls/e2e/e2e_test.go:235
#	0x53a6c3	testing.tRunner+0x163						/home/travis/.gimme/versions/go1.12.linux.amd64/src/testing/testing.go:865

1 @ 0x599d5e 0x599b2a 0x5959fc 0x738fd7 0x485a61
#	0x599d5d	runtime/pprof.writeRuntimeProfile+0x9d			/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/pprof/pprof.go:708
#	0x599b29	runtime/pprof.writeGoroutine+0xc9			/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/pprof/pprof.go:670
#	0x5959fb	runtime/pprof.(*Profile).WriteTo+0x4fb			/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/pprof/pprof.go:329
#	0x738fd6	github.com/pion/transport/test.TimeOut.func1+0x96	/home/travis/gopath/pkg/mod/github.com/pion/transport@v0.8.10/test/util.go:18

panic: timeout

goroutine 129 [running]:
github.com/pion/transport/test.TimeOut.func1()
	/home/travis/gopath/pkg/mod/github.com/pion/transport@v0.8.10/test/util.go:21 +0x164
created by time.goFunc
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/time/sleep.go:169 +0x52
FAIL	github.com/pion/dtls/v2/e2e	39.267s
```

### LINKED ISSUE #205: Refactor handshake state machine
ref: https://github.com/pion/dtls/issues/143

- Implement DTLS Timeout and Retransmission State Machine as specified in https://tools.ietf.org/html/rfc6347#page-22
  - Fix retransmission conditions
  - FIx handshake exit condition
- Refactor incoming handshake packet handling
  - Split all-in-one function to per-state function/interfaces

### PR DESCRIPTION
**Pull request by at-wat**
### Description

- Fully implement retransmition flow specified in RFC6347 Section 4.2.4.
  - parse incoming handshake message sequentially instead of calling type switch for each message
  - fix condition to end client handshake
  - return read error transparently from underlying connection
  - support packet contains both handshake and encrypted message (OpenSSL send such packet.)
- Some miscellaneous fixes
  - fix default ServerName on test
  - fix boundary check of ClientHello unmarshaller
  - verify epoch and CipherSuite
  - fix udp.Conn deadlock on Close()
  - fix routine leaks in tests 
### Reference issue
fixes #205 
fixes #143
fixes #145 (e52295e7d3eac0b1756c96a56a9a29add0014066)
fixes #178 (partially improved by c85233f9e892bdb74af36214e08f7e5b488ee2eb, fixed by 8d0a46be80271bf5599a0490a712220b165f3110)

### CODE REVIEW

**Comment by at-wat** [comment]
Flakiness of https://github.com/pion/dtls/issues/178 seems get worse. I'll dig into this bit more.

**Comment by at-wat** [comment]
State machine of RFC6347 Section 4.2.4 is implemented in [handshaker.go](https://github.com/pion/dtls/pull/208/files#diff-a7ca58a13dae66b30dd8b8a8db308ec7).
Parsing and generating each flight is implemented in [flight0handler.go](https://github.com/pion/dtls/pull/208/files#diff-31fad6b633f2caa4b33ff87112862d08) - [flight6handler.go](https://github.com/pion/dtls/pull/208/files#diff-b1341f474c2dcf90d0e3c5413640bff3).

**Comment by Sean-Der** [comment]
Really fantastic work @at-wat just starting to read through, but really love the small flight handlers :)

Also 0 API breakage AND you increased test coverage 🎆 

* Do you think it would be possible (or even worth it) to put the flighthandlers in an internal package? Maybe this isn't needed, but maybe it could make things easier to reason about. It could also be a huge pain since we share so much between everything :/

File: flight.go [hunk 1]
```diff
@@ -72,41 +66,20 @@ func (f flightVal) String() string {
...
-	if isClient {
-		val = flight1
-	}
-	return &flight{
-		val:           val,
-		workerTrigger: make(chan struct{}, 1),
-
-		log: logger,
+func (f flightVal) isLastSendFlight() bool {
+	switch f {
```
Comment (Sean-Der) [inline flight.go]: return `f == flight6`

File: flight.go [hunk 2]
```diff
@@ -72,41 +66,20 @@ func (f flightVal) String() string {
...
-func (f *flight) set(val flightVal) {
-	f.Lock()
-	f.log.Tracef("[handshake] Moving from %s to %s", f.val.String(), val.String())
-	f.val = val // TODO ensure no invalid transitions
-	f.Unlock()
-
-	select {
-	case f.workerTrigger <- struct{}{}:
+func (f flightVal) isLastRecvFlight() bool {
+	switch f {
```
Comment (Sean-Der) [inline flight.go]: `return f == flight5`

File: flighthandler.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	case flight3:
+		return flight3Generate
+	case flight4:
+		return flight4Generate
+	case flight5:
+		return flight5Generate
+	case flight6:
+		return flight6Generate
+	default:
+		return nil
```
Comment (Sean-Der) [inline flighthandler.go]: Unlikely to happen, but the caller doesn't handle `return nil` 

We should probably return an `error` from this function, and then will force the caller to handle the error (since tooling doesn't force us to think about `nil` return values)

File: flighthandler.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	case flight3:
+		return flight3Parse
+	case flight4:
+		return flight4Parse
+	case flight5:
+		return flight5Parse
+	case flight6:
+		return flight6Parse
+	default:
+		return nil
```
Comment (Sean-Der) [inline flighthandler.go]: https://github.com/pion/dtls/pull/208/files#r386566483

**Comment by at-wat** [comment]
> Do you think it would be possible (or even worth it) to put the flighthandlers in an internal package? Maybe this isn't needed, but maybe it could make things easier to reason about. It could also be a huge pain since we share so much between everything :/

`handshakeCache` and `State` should also be drastically refactored to split the package. 
I agree that doing it cleans the interface more.

**Comment by Sean-Der** [comment]
> > Do you think it would be possible (or even worth it) to put the flighthandlers in an internal package? Maybe this isn't needed, but maybe it could make things easier to reason about. It could also be a huge pain since we share so much between everything :/
> 
> `handshakeCache` and `State` should also be drastically refactored to split the package.
> I agree that doing it cleans the interface more.

:+1: I don't want to slow you down too much/make this PR super painful for you. So feel free to push back :)

Should this also solve https://github.com/pion/dtls/issues/145 ?

File: handshake_cache.go [hunk 5]
```diff
@@ -12,6 +14,7 @@ type handshakeCacheItem struct {
 type handshakeCachePullRule struct {
 	typ      handshakeType
 	isClient bool
+	optional bool
```
Comment (at-wat) [inline handshake_cache.go]: `encrypted bool` must be added to verify that Finished message is encrypted.

**Comment by at-wat** [comment]
`transport/test.bridgeConn` should support `net.Conn` `SetDeadline()` to suppress flaky test results.

File: flighthandler.go (hunk 3 above)
Comment (at-wat) [reply flighthandler.go]: Addressed your comments:+1:

**Comment by at-wat** [comment]
@Sean-Der Added some more fixes including #145. Some of the flaky tests were caused by routine leaks from other tests.

**Comment by at-wat** [comment]
TODO before merge:
- [ ] squash 5853912 and 3dd1747 into 4d5aeb6

File: conn.go [hunk 6]
```diff
@@ -32,71 +29,42 @@ var invalidKeyingLabels = map[string]bool{
...
+	encryptedPackets [][]byte
 
-	connectionClosed *closer.Closer // Closed on connection close and unblock read
-	handshakeErr     *atomicError   // Error if one occurred during handshake
-	readErr          *atomicError   // Error if one occurred in inboundLoop
+	connectionClosed       bool
+	connectionClosedByUser bool
+	closeLock              sync.Mutex
+	closed                 *closer.Closer
+	wg                     sync.WaitGroup
```
Comment (Sean-Der) [inline conn.go]: Mind naming `wg` something a little more self documenting `handshakeLoopsFinished`.

File: conn.go [hunk 7]
```diff
@@ -631,155 +543,202 @@ func (c *Conn) fragmentHandshake(h *handshake) ([][]byte, error) {
...
 		}
 
-		if err := c.handshakeErr.load(); err != nil {
-			return nil, err
-		}
-		if c.connectionClosed.Err() != nil {
-			return nil, io.EOF
-		}
+		c.closeLock.Lock()
+		connClosed := c.connectionClosed
```
Comment (Sean-Der) [inline conn.go]: Can you make `c.connectionClosed` an atomic, or a function that auto locks/unlocks? Juggling mutexes has ended poorly for us a few times, so just want to get away from that.

File: conn.go [hunk 8]
```diff
@@ -791,60 +750,137 @@ func (c *Conn) isHandshakeCompletedSuccessfully() bool {
...
-			if alertPtr != nil {
-				if alertErr := c.notify(alertPtr.alertLevel, alertPtr.alertDescription); alertErr != nil {
-					err = fmt.Errorf("%v %v", err, alertErr)
+		}
+	}()
+	go func() {
+		defer c.wg.Done()
+		for {
+			if err := c.readAndBuffer(ctxRead); err != nil {
+				switch e := err.(type) {
```
Comment (Sean-Der) [inline conn.go]: Not a big fan of unsetting a value on the stack. Would you mind refactoring this so it is more explicit? Maybe early returns are better?

File: conn.go [hunk 9]
```diff
@@ -791,60 +750,137 @@ func (c *Conn) isHandshakeCompletedSuccessfully() bool {
...
+	return err
+}
+
+func (c *Conn) close(byUser bool) error {
+	c.cancelHandshaker()
+	c.cancelHandshakeReader()
+	c.closed.Close()
+
+	if c.isHandshakeCompletedSuccessfully() && byUser {
+		_ = c.notify(context.Background(), alertLevelWarning, alertCloseNotify)
```
Comment (Sean-Der) [inline conn.go]: Are we throwing an error away here?

File: conn.go [hunk 10]
```diff
@@ -791,60 +750,137 @@ func (c *Conn) isHandshakeCompletedSuccessfully() bool {
...
 	}
 
-	c.workerTicker.Stop()
-	c.handshakeDoneSignal.Close()
-	c.connectionClosed.Close()
-	return c.nextConn.Close()
+	c.closeLock.Lock()
```
Comment (Sean-Der) [inline conn.go]: Getting/Setting things related to `close` should be in a function (or make everything atomic)

In the future developers will not know what is thread safe, could see races popping up accidentally.

File: flight0handler.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+import (
+	"context"
+	"crypto/rand"
+)
+
+func flight0Parse(ctx context.Context, c flightConn, state *State, cache *handshakeCache, cfg *handshakeConfig) (flightVal, *alert, error) {
+	// HelloVerifyRequest can be skipped by the server,
+	// so allow ServerHello during flight1 also
+	seq, msgs, ok := cache.fullPullMap(state.handshakeRecvSequence,
```
Comment (Sean-Der) [inline flight0handler.go]: I would put `!ok` first.

I know `fullPullMap` always gives a value valid for `seq`, but I needed to confirm. When I see `!ok` I assume all return values are no good.

Comment (Sean-Der) [reply flight0handler.go]: What do you think of moving the `flight requirements` out of the Parse functions?

I am sure it is fine, but seems fragile to do the checking/asserting in every function. Would be nice if each Parse function was called with everything it needed.

My vote is that we fix this in a follow up PR, but merge what you have now :)

Comment (Sean-Der) [reply flight0handler.go]: I would just be really nice if these functions weren't re-entrant.

File: conn.go (hunk 7 above)
Comment (at-wat) [reply conn.go]: Read/write of `connectionClosedByUser` and `connectionClosed` must be synchronized at https://github.com/pion/dtls/pull/208/files#r388001174. Otherwise, it causes race condition.

Standard `net.Conn` implementations like `TCPConn` and `UDPConn` seems doesn't returning error from `Close()` at the first time of the connection is internally closed. This part is a logic for it.

File: fragment_buffer_test.go [hunk 12]
```diff
@@ -47,12 +47,12 @@ func TestFragmentBuffer(t *testing.T) {
 			}
 		}
 
-		out := fragmentBuffer.pop()
+		out, _ := fragmentBuffer.pop()
```
Comment (Sean-Der) [inline fragment_buffer_test.go]: `assert.NoError(t, err)`

File: fragment_buffer_test.go [hunk 13]
```diff
@@ -47,12 +47,12 @@ func TestFragmentBuffer(t *testing.T) {
...
 		if !reflect.DeepEqual(out, test.Expected) {
 			t.Errorf("fragmentBuffer '%s' push/pop: got % 02x, want % 02x", test.Name, out, test.Expected)
 		}
 
-		if frag := fragmentBuffer.pop(); frag != nil {
+		if frag, _ := fragmentBuffer.pop(); frag != nil {
```
Comment (Sean-Der) [inline fragment_buffer_test.go]: `assert.NoError(t, err)`

File: fragment_buffer_test.go (hunk 12 above)
Comment (Sean-Der) [inline fragment_buffer_test.go]: Can you assert the `uint16` as well, just to make sure that doesn't regress/break.

**Comment by Sean-Der** [comment]
Really amazing work @at-wat. This is a VERY welcome change, I am really impressed. I feel pretty confident about this change because our coverage is good, but there is too much for me to understand/review completely :)

In a follow up commit I think it would be really cool if we could.
* Move all marshal/unmarshal of handshake messages to another package
* Move all state machine logic to another package

Maybe I am over complicating things, but it would be nice to break things into more manageable chunks.

File: flight0handler.go (hunk 11 above)
Comment (at-wat) [reply flight0handler.go]: > What do you think of moving the flight requirements out of the Parse functions?

One exception making it complexed is `flight4Parse`.
Flight4 contains `Finished` message, that it is not available to be decrypted until parsing `ClientKeyExchange`.

File: conn.go (hunk 7 above)
Comment (at-wat) [reply conn.go]: Adding `isConnectionClosed()` helper and some more comment.

File: conn.go (hunk 9 above)
Comment (at-wat) [reply conn.go]: This is called also on already closed `nextConn`.
Even in such case, I would like to keep returning non-error on the first user call of `Close()`.

**Comment by at-wat** [comment]
@Sean-Der Addressed your comments!

File: fragment_buffer_test.go (hunk 12 above)
Comment (at-wat) [reply fragment_buffer_test.go]: Added an assert of returned epoch and an another test data having non-zero epoch.

File: conn.go (hunk 9 above)
Comment (at-wat) [reply conn.go]: Added a comment about this.

File: conn.go (hunk 8 above)
Comment (at-wat) [reply conn.go]: Updated to use `continue` instead of `err = nil`.

**Comment by at-wat** [comment]
squashed some commits

**Comment by Sean-Der** [comment]
@at-wat I am really happy with this branch. IMO I am happy with merging this into master! If you want to add other reviewers that also could be a helpful.

This is a huge step in the right direction :) excited to see all those long standing bugs squashed.

**Comment by at-wat** [comment]
Thanks. Let me merge for now.
If anyone find problems, please add comments here or directly fix them.
